src/ugeneui/src/main_window/CreateDesktopShortcutTask.cpp

```diff
@@ -0,0 +1,193 @@
+                MultiByteToWideChar(CP_ACP, 0, pathLink, -1, wsz, MAX_PATH);
+
+                // Add code here to check return value from MultiByteWideChar
+                // for success.
+
+                // Save the link by calling IPersistFile::Save.
+                hres = ppf->Save(wsz, TRUE);
+                ppf->Release();
+            } else {
+                return false;
+            }
+        }
+        psl->Release();
+    }
+    return SUCCEEDED(hres);
+#elif defined(Q_OS_LINUX)
+    QString homeDir = QDir::homePath();
+    QFile link(homeDir + "/Desktop/UGENE.desktop");
+    if (link.open(QFile::WriteOnly | QFile::Truncate)) {
+        QTextStream out(&link);
+        out.setCodec("UTF-8");
+        out << "[Desktop Entry]" << endl
+            << "Encoding=UTF-8" << endl
+            << "Version=1.0" << endl
+            << "Type=Application" << endl
+            << "Terminal=false" << endl
+            << "Exec=" + QCoreApplication::applicationFilePath() << endl
+            << "Name=Unipro UGENE" << endl
+            << "Icon=" + QCoreApplication::applicationDirPath() + "/ugene.png" << endl
+            << "Name[en_US]=Unipro UGENE" << endl;
+        link.close();
+        if (!link.setPermissions(link.permissions() | QFileDevice::ExeOwner | QFileDevice::ExeUser)) {
+            return false;
+        }
+        return true;
+    }
+    return false;
+#elif defined(Q_OS_MACX)
+    return false;
+#endif    // Q_OS_WIN
+}
+
+Task::ReportResult CreateDesktopShortcutTask::report() {
+    if (hasError() || startError) {
+        return ReportResult_Finished;
+    }
+
+    Version thisVersion = Version::appVersion();
+
+    Answer answer = DoNothing;
+    if (runOnStartup) {
+    } else {
+        DesktopShortcutMessage message("xxxx");
+        answer = message.getAnswer();
+    }
+
+    switch (answer) {
+    case Create:
+        createDesktopShortcut();
+        break;
+    case DoNothing:
+    default:
+        break;
+    }
+    return ReportResult_Finished;
+}
+
+void CreateDesktopShortcutTask::sl_registerInTaskScheduler() {
+    AppContext::getTaskScheduler()->registerTopLevelTask(this);
+}
+
+/************************************************************************/
+/* UpdateMessage */
+/************************************************************************/
+DesktopShortcutMessage::DesktopShortcutMessage(const QString &newVersion) {
+    QWidget *parent = AppContext::getMainWindow()->getQMainWindow();
+    const QString message = tr("Create desktop shortcut?").arg(newVersion);
+    const QString title = tr("Desktop shortcut");
+
+    dialog = new QMessageBox(QMessageBox::Question, title, message, QMessageBox::NoButton, parent);
+
+    createButton = dialog->addButton(tr("Create"), QMessageBox::ActionRole);
+    dialog->addButton(tr("Cancel"), QMessageBox::ActionRole);
+}
+
+CreateDesktopShortcutTask::Answer DesktopShortcutMessage::getAnswer() const {
+    dialog->exec();

```

Comment:
Why don't you create the dialog here or just use the static method of QMessageBox?

Reply:
sure, done